ssl/time.c

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <errno.h>
+#include <openssl/err.h>
+#include "internal/time.h"
+
+OSSL_TIME ossl_time_now(void)
+{
+#if defined(_WIN32)
+    SYSTEMTIME st;
+    union {
+        unsigned __int64 ul;
+        FILETIME ft;
+    } now;
+
+    GetSystemTime(&st);
+    SystemTimeToFileTime(&st, &now.ft);
+    /* re-bias to 1/1/1970 */
+# ifdef  __MINGW32__
+    now.ul -= 116444736000000000ULL;
+# else
+    now.ul -= 116444736000000000UI64;
+# endif
+    return ((uint64_t)now.ul) * (OSSL_TIME_SECOND / 10000000);
+#else
+    struct timeval t;
+
+    if (gettimeofday(&t, NULL) < 0) {
```

Comment:
This is `CLOCK_REALTIME` so it can jump around, therefore we should not use it to schedule QUIC timer events. `ifdef CLOCK_MONOTONIC` is probably a decent way to detect if the platform supports `clock_gettime`.  `gettimeofday` might be a suboptimal but tolerable fallback if it's not available.

The Windows equivalent you want is `QueryPerformanceCounter` (which is what C++'s `std::chrono::steady_clock` uses). For MacOS, `clock_gettime(CLOCK_UPTIME_RAW)` apparently. Source: https://github.com/ros2/rcutils/issues/43

Reply:
`CLOCK_MONOTONIC_RAW` would be ideal I think.  The approach I've used in the past is a series of preprocessor checks for the various clocks in preferential order:
```c
#if defined(CLOCK_MONOTONIC_RAW)
#elif defined(CLOCK_MONOTONIC)
#elif defined(CLOCK_REALTIME)
#else
   // gettimeofday
#endif
```
Usually with some OS specific interventions because everyone does this slightly differently.

I think the timer used should be a separate discussion probably involving a long chunk of OTC time.

Reply:
Refer to #18519 for the decision.  I'll mark this as resolved since I think the specific choice is separate.